### DISCUSSION TITLE
Question: How can I reference requirements in implementations?

### DISCUSSION THREAD

**Discussion by fkromer** [strictdoc-project/strictdoc#569] [Q&A] [answered]
As a developer I would like to be able to reference requirements in the implementation of production and test code (e.g. via UID in doxygen/docstring). Is this possible already?

**Comment by stanislaw** [comment]
This feature has been implemented a while ago, but I am pretty sure it has not received enough testing. Stay warned of possible issues and rough edges.

I am finishing the initial ReqIF support and the plan is to release the reqs-files traceability and ReqIF documentation in a same release (probably it's time for a good 0.1 version).

For now, you can check the integration tests for this feature here: `tests/integration/commands/export/html/file_traceability`.

The command that you use in that case is `strictdoc export ... --experimental-enable-file-traceability`, see the tests.

**Reply by fkromer** [reply]
> This feature has been implemented a while ago, but I am pretty sure it has not received enough testing. Stay warned of possible issues and rough edges.

[awslabs/duvet](https://github.com/awslabs/duvet) is used in the security context and uses duvet-core. As soon as [duvet-core](https://lib.rs/crates/duvet-core) (Rust) provides a stable API it could be used in Strictdoc as a programming language generic requirement and test UID parsing backend (provided in source code comments). The Rust lib could be wrapped into a Python library maintained in the duvet project (preferred) or strictdoc project. Could be worth reaching out to @camshaft for more information.

**Reply by stanislaw** [reply, marked as answer]
Hi @fkromer,

StrictDoc can do a general parsing of text files with @relation markers for quite a while now and as of 2024-Q4 it can parse Python/C/C++ in terms of functions and classes.

https://strictdoc.readthedocs.io/en/latest/latest/docs/strictdoc_01_user_guide.html#SECTION-TRACEABILITY-REQS-TO-SOURCE-CODE

(Thanks for sharing the link about duvet, I didn't know about it.)

**Reply by fkromer** [reply]
Cross-language support is getting more and more important in the security context. In contrast to safety system requirements security system child requirements are usually implemented vertically across the tech stack. Rust is getting more and more common in the (embedded) backend. Typescript is pretty common in the frontend. Im distributed systems it‘s even common to consider Kotlin (multiplatform), Swift, … as well.

**Comment by fkromer** [comment]
> I am finishing the initial ReqIF support and the plan is to release the reqs-files traceability and ReqIF documentation in a same release (probably it's time for a good 0.1 version).

Are there other requirement interchange formats in addition to [ReqIF](https://www.omg.org/spec/ReqIF) btw?

**Comment by stanislaw** [comment]
> > I am finishing the initial ReqIF support and the plan is to release the reqs-files traceability and ReqIF documentation in a same release (probably it's time for a good 0.1 version).
> 
> Are there other requirement interchange formats in addition to [ReqIF](https://www.omg.org/spec/ReqIF) btw?

Not that I am aware of.

I have only seen this one https://github.com/doorstop-dev/doorstop/issues/507 but have never looked into its details.

**Comment by stanislaw** [comment]
StrictDoc supports connecting requirements and implementations tests. See the documentation here: https://strictdoc.readthedocs.io/en/latest/latest/docs/strictdoc_01_user_guide.html#SECTION-TRACEABILITY-REQS-TO-SOURCE-CODE.

Somehow I lost track of this issue. I think it can now be closed properly.
